sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala

```diff
@@ -704,6 +704,23 @@ class PlannerSuite extends SharedSQLContext {
     df.queryExecution.executedPlan.execute()
   }
 
+  test("SPARK-25278: physical nodes should be different instances for same logical nodes") {
+    val range = Range(1, 1, 1, 1)
+    val df = Union(range, range)
+    val ranges = df.queryExecution.optimizedPlan.collect {
+      case r: Range => r
+    }
+    assert(ranges.length == 2)
+    // Ensure the two Range instances are equal according to their equal method
+    assert(ranges.head == ranges.last)
+    val execRanges = df.queryExecution.sparkPlan.collect {
+      case r: RangeExec => r
+    }
+    assert(execRanges.length == 2)
+    // Ensure the two RangeExec instances are different instances
+    assert(!execRanges.head.eq(execRanges.last))
```

Comment:
Is it difficult to test the `eq` behaivour by using the `QueryPlanner` class?

Reply:
mmmh, what do you mean exactly?

Reply:
e.g.,
```
case class DummyPlanner() extends QueryPlanner[LogicalPlan] { ... }

assert(DummyPlanner().plan(eqTestPlan) === expectedPlan))
```

Reply:
I think it is doable, but I don't see the advantage. May you please explain me why that is better? Thanks.

Reply:
just a question we can do that. The current test is ok to me.

Reply:
cool, thanks. Anyway, yes, I think we can do that. Thanks.